aria_plugin/operations.py

```diff
@@ -0,0 +1,138 @@
+from cloudify.decorators import operation
+
+
+@operation
+def create():
+
+    _mk_working_dir()
+    _init_model_storage()
+    _init_resource_storage()
+
+    _generate_csar_path()
+    _extract_csar()
+
+    _prepare_plugins_for_installation()
+    _install_plugins()
+    _cleanup_csar_plugins()
+
+    _store_service_template()
+
+    _create_service()
+
+
+def _create_if_not_existing(path):
+    try:
+        os.makedirs(path)
+    except OSError as e:
+        if e.errno != errno.EEXIST:
+            raise
+    return path
+
+
+def _mk_working_dir():
+    dir_name = 'aria-{tenant_name}'.format(tenant_name=ctx.tenant_name)
+    abs_path = os.path.join(env.cloudify_plugins_dir, dir_name)
+    env._workdir = _create_if_not_existing(abs_path)
+
+    # create subdirectories
+    _create_if_not_existing(env.aria_plugins_dir)
+    _create_if_not_existing(env.model_storage_dir)
+    _create_if_not_existing(env.resource_storage_dir)
+
+
+def _init_model_storage():
+    initiator_kwargs = {'base_dir': env.model_storage_dir}
+    env.model_storage = aria.application_model_storage(
+        api=SQLAlchemyModelAPI, initiator_kwargs=initiator_kwargs)
+
+
+def _init_resource_storage():
+    api_kwargs = {'directory': env.resource_storage_dir}
+    env.resource_storage = aria.application_resource_storage(
+        api=FileSystemResourceAPI, api_kwargs=api_kwargs)
+
+
+def _generate_csar_path():
+    csar_path = ctx.node.properties[CSAR_PATH_PROPERTY]
+    if '://' not in csar_path:
+        # the csar_path property is relative to the blueprint's directory
+        csar_path = os.path.join(env.blueprints_dir, ctx.tenant_name,
+                                 ctx.blueprint.id, csar_path)
+    env.csar_source = csar_path
+
+
+def _extract_csar():
+    env.tmp_csar_dest = tempfile.mkdtemp(prefix='tmp-csar-')
+    env.tmp_files.append(env.tmp_csar_dest)
+    env.csar = csar.read(source=env.csar_source, destination=env.tmp_csar_dest)
+
+
+def _prepare_plugins_for_installation():
+    plugins = set(os.listdir(env.csar_plugins_dir))
+    plugins_to_install = set(ctx.node.properties[PLUGINS_PROPERTY])
+    missing_plugins = plugins_to_install - plugins
+
+    if missing_plugins:
+        raise PluginsException('Requested plugins {0} is not in the csar '
+                               '`plugins` directory'
+                               .format(', '.join(missing_plugins)))
+    else:
+        env.plugins_to_install = ['{name}{ext}'.format(name=plugin_name,
+                                                       ext=WAGON_EXTENSION)
+                                  for plugin_name in plugins_to_install]
+
+
+def _install_plugins():
+    env.plugin_manager = PluginManager(model=env.model_storage,
+                                       plugins_dir=env.aria_plugins_dir)
+    for plugin_to_install in env.plugins_to_install:
+        plugin_path = os.path.join(env.csar_plugins_dir, plugin_to_install)
+        env.plugin_manager.validate_plugin(plugin_path)
```

Comment:
check if install already validates

Reply:
In the CLI it does, but not in the via the `PluginManager` itself.